### PR TITLE
Refactor llcp pdu length validation

### PR DESCRIPTION
**Pull request by erbr-ot**
Using a macro instead of repeated code for length validation

### CODE REVIEW

**Comment by erbr-ot** [comment]
sits on-top of SCA PR, will rebase once merged

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 1]
```diff
@@ -1178,290 +1210,192 @@ static bool pdu_is_terminate(struct pdu_data *pdu)
 	return pdu->llctrl.opcode == PDU_DATA_LLCTRL_TYPE_TERMINATE_IND;
 }
 
+#define VALIDATE_PDU_LEN(pdu, type) (pdu->len == sizeof(pdu->llctrl.type) + 1U)
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:1213]:
```suggestion
#define VALIDATE_PDU_LEN(pdu, type) (pdu->len == PDU_DATA_LLCTRL_LEN(type))
```

**Comment by Thalley** [comment]
> sits on-top of SCA PR, will rebase once merged

Can you please provide at link to the other PR? Also suggest to mark this as draft until the other PR has been merged

**Comment by hermabe** [comment]
> > sits on-top of SCA PR, will rebase once merged
> 
> Can you please provide at link to the other PR? Also suggest to mark this as draft until the other PR has been merged

It is also possible to set the target branch of this PR to the branch of the other PR. Then the diff will show up correctly and once the other PR is merged the target branch of this PR will be changed to main.

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 1 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: Ack

**Review by Thalley — DISMISSED** [review]

**Review by thoh-ot — CHANGES_REQUESTED** [review]
Change is fine for non-test code, but for tests/bluetooth/controller/common/src/helper_pdu.c I'm a bit "worried" that this change uses `PDU_DATA_LLCTRL_LEN` for both the `helper_pdu_encode_XXX` functions and the `helper_pdu_verify_XXX` functions.

If an error was to be introduced in `PDU_DATA_LLCTRL_LEN`, these tests would no longer catch the length errors - as it's a fairly simple macro the chances are pretty low - I'm thinking in broader terms.

In my world test code is close to the metal and sometimes at bit cumbersome due to the fact that it cannot rely on the constructs it's trying to test and verify - otherwise there's no point in testing if both ends of the test has the same error and therefore self corrects.

Would like to hear other opinions.

**Comment by Thalley** [comment]
> Change is fine for non-test code, but for tests/bluetooth/controller/common/src/helper_pdu.c I'm a bit "worried" that this change uses `PDU_DATA_LLCTRL_LEN` for both the `helper_pdu_encode_XXX` functions and the `helper_pdu_verify_XXX` functions.
> 
> If an error was to be introduced in `PDU_DATA_LLCTRL_LEN`, these tests would no longer catch the length errors - as it's a fairly simple macro the chances are pretty low - I'm thinking in broader terms.
> 
> In my world test code is close to the metal and sometimes at bit cumbersome due to the fact that it cannot rely on the constructs it's trying to test and verify - otherwise there's no point in testing if both ends of the test has the same error and therefore self corrects.
> 
> Would like to hear other opinions.

If I am understanding you correct, you are arguing that using the same function/macro/whatever to calculate the value and to verify that the calculation is correct, may not catch any issues. I can fully agree to that, basically you don't want to have a test like

```c
int foo()
{
    return val;
}

int test_foo()
{
    foo_val = foo();
    zassert_equal(foo_val, foo());
}
```

As any errors introduced in `foo` would not be caught.

**Comment by erbr-ot** [comment]
I agree with @thoh-ot and @Thalley  - updated PR to keep the test code as is.

**Review by Thalley — DISMISSED** [review]

**Comment by cvinayak** [comment]
Added DNM as there is only one squashed commit in this PR. Please rebase.

**Comment by thoh-ot** [comment]
> If I am understanding you correct, you are arguing that using the same function/macro/whatever to calculate the value and to verify that the calculation is correct, may not catch any issues. I can fully agree to that, basically you don't want to have a test like...

@Thalley That was exactly what i was trying to say.

**Review by thoh-ot — APPROVED** [review]

**Review by Thalley — APPROVED** [review]

**Review by cvinayak — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Pull request merged**